### PR TITLE
added ability to define which utxos to use

### LINKED ISSUE #44: Manual UTXO selection
Some covenant contracts are reliant on having specific UTXOs as inputs. It should be possible to optionally provide the UTXOs to the transaction manually. For this it should also be possible to retrieve all UTXOs that belong to a contract instance (`instance.getUTXOs()`).

### PR DESCRIPTION
**Pull request by maikelmclauflin**
fixes #44

### CODE REVIEW

File: packages/cashscript/src/Transaction.ts [hunk 1]
```diff
@@ -115,7 +115,13 @@ export class Transaction {
 
     this.builder.setLockTime(locktime);
 
-    const { inputs, outputs } = await this.getInputsAndOutputs(outs, options?.fee);
+    const { inputs, outputs } = await this.getInputsAndOutputs(
+      outs,
+      options?.fee,
+      DUST_LIMIT,
```
Comment (rkalis) [inline packages/cashscript/src/Transaction.ts]: It looks like I made a mistake in here. It should pass in `options?.minChange` here, but it looks like I forgot to do this in `0.3.2`. Could you replace `DUST_LIMIT` with `options?.minChange`?

File: packages/cashscript/test/e2e/P2PKH.test.ts [hunk 2]
```diff
@@ -50,6 +54,48 @@ describe('P2PKH', () => {
       const txOutputs = getTxOutputs(tx);
       expect(txOutputs).toEqual(expect.arrayContaining([{ to, amount }]));
     });
+
+    it('should succeed when defining its own utxos', async () => {
```
Comment (rkalis) [inline packages/cashscript/test/e2e/P2PKH.test.ts]: This test looks nice. But could you split it up into two separate 'should's? One for the failure (`tooHigh`), and one for the success?

File: packages/cashscript/src/Transaction.ts [hunk 3]
```diff
@@ -169,8 +175,12 @@ export class Transaction {
     hardcodedFee?: number,
     minChange: number = DUST_LIMIT,
     satsPerByte: number = 1.0,
+    _utxos?: Utxo[],
   ): Promise<{ inputs: Utxo[], outputs: OutputForBuilder[] }> {
-    const { utxos } = await this.bitbox.Address.utxo(this.address) as AddressUtxoResult;
+    let utxos = _utxos;
```
Comment (rkalis) [inline packages/cashscript/src/Transaction.ts]: So when someone passes in their own list of inputs, all of them should be used. I think that with this code change, it only replaces the "pool" of possible UTXOs, but then it still performs the UTXO selection algorithm below. This means that some of the manual inputs might get excluded due to the selection process. 

So instead, if a manual list of utxos is provided, this should completely replace the "selection process". But it should still add a change output where necessary, and it should also check that the provided UTXOs are enough to cover the transaction. Does that make sense?

File: packages/cashscript/test/Instance.test.ts [hunk 4]
```diff
@@ -27,6 +30,19 @@ describe('Instance', () => {
     });
   });
 
+  describe('Contract Utxo inputs', () => {
```
Comment (rkalis) [inline packages/cashscript/test/Instance.test.ts]: It's difficult to test this really well so this is good 👍

File: packages/cashscript/src/Instance.ts [hunk 5]
```diff
@@ -29,6 +43,28 @@ export class Instance {
     return details.balanceSat + details.unconfirmedBalanceSat;
   }
 
+  async getUnconfirmed(): Promise<UnconfirmedUtxo[]> {
```
Comment (rkalis) [inline packages/cashscript/src/Instance.ts]: Could you make this private?

File: packages/cashscript/src/Instance.ts [hunk 6]
```diff
@@ -29,6 +43,28 @@ export class Instance {
...
+    const { utxos } = await this.bitbox.Address
+      .unconfirmed(this.address) as AddressUnconfirmedResult;
+    return utxos;
+  }
+
+  async getUtxo(): Promise<Utxo[]> {
```
Comment (rkalis) [inline packages/cashscript/src/Instance.ts]: Could you make this private?

File: packages/cashscript/src/Transaction.ts (hunk 1 above)
Comment (maikelmclauflin) [reply packages/cashscript/src/Transaction.ts]: makes sense. happy to

File: packages/cashscript/src/Transaction.ts [hunk 7]
```diff
@@ -169,8 +175,12 @@ export class Transaction {
     hardcodedFee?: number,
     minChange: number = DUST_LIMIT,
     satsPerByte: number = 1.0,
+    _utxos?: Utxo[],
```
[the remark this answers is not in the document]
Comment (maikelmclauflin) [reply packages/cashscript/src/Transaction.ts]: can do

File: packages/cashscript/src/interfaces.ts [hunk 8]
```diff
@@ -7,6 +7,18 @@ export interface Utxo {
...
+  vout: number;
+  scriptPubKey: string;
+  amount: number;
+  satoshis: number;
+  confirmations: number;
+  ts: number;
+  legacyAddress: string;
+  cashAddress: string;
+}
+
```
[the remark this answers is not in the document]
Comment (maikelmclauflin) [reply packages/cashscript/src/interfaces.ts]: make sense, can do

File: packages/cashscript/test/Instance.test.ts (hunk 4 above)
Comment (maikelmclauflin) [reply packages/cashscript/test/Instance.test.ts]: removed as private methods prevent testing

**Comment by rkalis** [comment]
I'm planning to change the Transaction send interface around, so I'll merge this change in, and make some changes to fit in the new interface.

**Pull request merged**